Chemmaster reagent transfer runtime fix

zoomachina opened the pull request:
## About The Pull Request
Fixes #84938 by adding a check for whether a beaker is inserted into the machine, and added a feedback message.
## Why It's Good For The Game
Fixes #84938.
## Changelog
:cl:
fix: deleting reagents from the chemmaster buffer works without a beaker inserted
/:cl:

zoomachina committed:
beaker? no this is patrick

zoomachina committed:
is_transfering -> transfer_mode
isTransfering -> transferMode

zoomachina committed:
changed feedback messages to be prettier because these aren't balloon alerts

LemonInTheDark commented on code/modules/reagents/chemistry/machinery/chem_master.dm:
```diff
@@ -425,21 +425,24 @@
 			if(target == "buffer")
-				return transfer_reagent(ui.user, beaker.reagents, reagents, reagent_ref, amount, TRUE)
+				return transfer_reagent(ui.user, beaker?.reagents, reagents, reagent_ref, amount, should_transfer)
 			else if(target == "beaker")
-				return transfer_reagent(ui.user, reagents, beaker.reagents, reagent_ref, amount, is_transfering)
+				return transfer_reagent(ui.user, reagents, beaker?.reagents, reagent_ref, amount, should_transfer)
```
this code seems really weird to me. I don't like the ?.reagents bit since it just silences a potential runtime, but then looking at it if should_transfer is FALSE it's valid?

transfer_reagent as a proc is doing too much shit. If we're trying to modularize we should split it up into code that asks for an amount, consumes power, and then use those procs to rebuild behavior here. at least IMO. this is maybe not the best solution but having args that only sometimes do anything is one hell of a code smell. I'd like to fix the structure here rather then just a symptom of the structure.

zoomachina replied:
yea I figured that you don't actually need `beaker.reagents` when the task is to delete reagents from the buffer. I wanted to make a separate proc for that case, but then I looked at `transfer_reagent` itself and found out that most of it is just sanity checks & the actual functionality is like 3 lines of code which call some other procs, so my options for changing the structure were 
1) move all the sanity checks to `ui_act` which seemed bad,
2) separate them into another proc (which would practically lead to the same structure as before, just more bloated) or 
3) make 2 procs with 80% of them being the same checks, which smelled just as bad.
In the end, I figured this is the better solution, maybe I'm missing something but I just don't know how else I can change the structure of this thing

zoomachina replied:
that said I think I can at least clean up `transfer_reagent` a little

zoomachina committed:
moved setting custom amount to ui_act, made transfer_reagent call a little cleaner I think

zoomachina commented:
I decided to move setting the custom amount to `ui_act` because I feel like it's better if `transfer_reagent` doesn't call any tgui things, as a result it doesn't need the `user` parameter anymore so I removed it.
To make calling `transfer_reagent` not smell as much, I changed it up a little, so that it's more explicit how it works. I tested it and it works fine, so I think this is ok?

zoomachina committed:
remove that since it's checked later anyway

zoomachina committed:
eh we don't really need this I guess

zoomachina commented:
I'm kind of confused by this check:
```
if(!use_energy(active_power_usage, force = FALSE))
		return FALSE
```
The idea seems elementary - if there's no power then the operation shouldn't commence, if there is we consume some energy from the APC. But as is, `transfer_reagent` can fail because of some other check after this power consumption, in which case the energy gets spent for no reason. Therefore, this code should be the last thing before the operation itself, which is ok I guess, but it means that I cannot move it to `ui_act`.

Another thing is, `active_power_usage` is the 'static power load on the APC' according to the comments. But here, it's used like the amount of energy spent every time this operation occurs. Is this even supposed to happen? It's not written in the comments, but it seems like `use_energy` is supposed to accept joules, while `active_power_usage` is watts. I remember reading that joules and watts are distinct, so this piece of code leaves too many questions and not enough answers to me. I'll try to solve this mystery, but in the meantime I'm drafting the PR

SyncIt21 commented:
> Another thing is, active_power_usage is the 'static power load on the APC' according to the comments. But here, it's used like the amount of energy spent every time this operation occurs. Is this even supposed to happen?

Yes. The comment about active power usage is outdated, this pattern of using power is everywhere  .
Anytime you want your machine to use power for a user driven operation(something started in the UI) then we use active power usage this way.

Joules and watts only differ when there is seconds_per_tick involved else they have the same value so there's no difference here.

In the meantime please address https://github.com/tgstation/tgstation/pull/84939#discussion_r1676837944

zoomachina committed:
revert is_transfering rename

zoomachina commented:
> > Another thing is, active_power_usage is the 'static power load on the APC' according to the comments. But here, it's used like the amount of energy spent every time this operation occurs. Is this even supposed to happen?
> 
> Yes. The comment about active power usage is outdated, this pattern of using power is everywhere.
> 
> Anytime you want your machine to use power for a user driven operation(something started in the UI) then we use active power usage this way.
> 
> Joules and watts only differ when there is seconds_per_tick involved like inside your process proc.In that case we convert power(Watts) to energy(Joules) and then pass it to use_energy() else they have the same value so there's no difference here.

I'll just leave it as is then

Ghommie commented on code/modules/reagents/chemistry/machinery/chem_master.dm:
```diff
@@ -425,18 +411,31 @@
 			var/target = params["target"]
 
+			if(amount == -1) // Set custom amount
+				amount = tgui_input_number(ui.user, "Enter amount to transfer", "Transfer amount")
```
```suggestion
				var/mob/user = ui.user //Hold a reference to the user if the UI is closed
				amount = tgui_input_number(user, "Enter amount to transfer", "Transfer amount")
				if(!amount || !user.can_perform_action(src))
					return FALSE
```

zoomachina replied:
> `if(!amount)`

it's checked later in transfer_reagents proc but I suppose theres' no harm in duplicating that check

zoomachina committed:
Apply suggestions from code review

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>
